server/src/main/scala/Actors.scala

```diff
@@ -84,36 +91,216 @@ trait Actors extends DynamoDb { self: core.Core with core.Configuration with git
       case IssueCommentEvent("created", issue, comment, _) =>
         prActor(issue.number) ! comment
 
-      case js@JobState(_, _, bs) =>
-        (for {
-          prParam <- Try(bs.parameters(PARAM_PR)) // we only care about jobs we started, and which thus have this parameter (when restarted manually, they should be carried forward automatically)
-          prNum   <- Try(prParam.toInt)
-        } prActor(prNum) ! js) // TODO: report failure
+      case PushEvent(ref, commits, _) =>
+        for {
+          ms  <- milestoneForBranch(ref) // only build for master-like branches (we always have milestones that mention this ref for those)
+          res <- buildPushedCommits(new BaseRef(ref), commits)
+        } yield res
+
+      // there are two cases, unfortunately:
+      //   - PARAM_PR's value is an integer ==> assumed to be PR number,
+      //   - otherwise, it's assumed to be the branch name of a job launched for a pushed commit (e.g., a merge or a direct push)
+      // TODO: report failure
+      case js@JobState(name, _, BuildState(number, _, _, _, _, _, _)) =>
```

Comment:
I liked your original suggestion better. Foreach is the right combinator here, no need to flatMap